### PR TITLE
Refactor VZV code to center on (and Deref to) VarZeroSlice; make VZVBorrowed private

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes https://github.com/unicode-org/icu4x/issues/1367

VZV and VZVOowned now deref to VZSlice, and most functionality is now on VZSlice. This simplifies our code greatly.

VZVBorrowed has been renamed to VZVComponents and is now crate-private; it still needs to exist to hold all of the unsafe parsed representation code.

As mentioned in https://github.com/unicode-org/icu4x/issues/1367#issuecomment-997338820 , there was no appreciable difference in the benchmarks.

A thing I didn't do here, but could have, is marking the `varzerovec` module as private and reexporting VZVOwned up top. Right now we have ZV, VZV, ZS, VZS up top which is nice, but now that there's only one other type in `varzerovec` it might be worth reexporting it too so that there's no extra module. It doesn't sound like a huge deal either way.

### CODE REVIEW

**Comment by sffc** [comment]
I think one of my PRs caused you merge conflicts ☹️

**Comment by Manishearth** [comment]
Argh, you named get_encoded_slice to as_encoded_bytes, I named it to as_bytes, this is going to be a mess

**Comment by Manishearth** [comment]
Do you think we should stick ot as_bytes here or move it over to into_encoded_bytes?

File: utils/zerovec/src/varzerovec/slice.rs [hunk 1]
```diff
@@ -60,41 +61,136 @@ pub struct VarZeroSlice<T: ?Sized> {
...
-    /// Obtain a [`VarZeroVecBorrowed`] borrowing from the internal buffer
+    /// Construct a new empty VarZeroSlice
+    pub fn new_empty() -> &'static Self {
+        let arr: &[u8] = &[];
+        unsafe { mem::transmute(arr) }
+    }
+    /// Obtain a [`VarZeroVecComponents`] borrowing from the internal buffer
     #[inline]
-    pub fn as_borrowed<'a>(&'a self) -> VarZeroVecBorrowed<'a, T> {
+    pub(crate) fn as_components<'a>(&'a self) -> VarZeroVecComponents<'a, T> {
```
Comment (sffc) [inline utils/zerovec/src/varzerovec/slice.rs:71]: Thought: This is a hot function, and I'm unsure about the code size implications.  If everything gets super inlined, the `VarZeroVecComponents` need not ever actually exist, and overall code size might co down.  But if these functions are never inlined, then we can do more code sharing.

**Comment by Manishearth** [comment]
I'm going to go ahead with the rename in this PR because it does more and does it consistently, can undo in a subsequent commit

File: utils/zerovec/src/varzerovec/slice.rs [hunk 2]
```diff
@@ -60,41 +61,136 @@ pub struct VarZeroSlice<T: ?Sized> {
...
+    /// let strings = vec!["foo", "bar", "baz", "quux"];
+    /// let bytes = VarZeroVec::<str>::get_serializable_bytes(&strings).unwrap();
+    ///
+    /// let mut vec: VarZeroVec<str> = VarZeroVec::parse_byte_slice(&bytes)?;
+    /// assert_eq!(vec.len(), 4);
+    /// # Ok::<(), ZeroVecError>(())
+    /// ```
     pub fn len(&self) -> usize {
-        self.as_borrowed().len()
+        self.as_components().len()
```
Comment (sffc) [inline utils/zerovec/src/varzerovec/slice.rs:95]: Thought: I was thinking that we would write custom code for each of these functions.  To get the length, we just read the first 4 bytes.  In theory, the compiler should figure that out, iff everything is inlined and DCE works fully.  But I don't know if I want to rely on that.

**Review by sffc — COMMENTED** [review]
Not approving or requesting changes yet, but opening some discussion threads.

File: utils/zerovec/src/varzerovec/slice.rs (hunk 2 above)
Comment (Manishearth) [reply utils/zerovec/src/varzerovec/slice.rs:94]: I kinda like that all the implementation code is in the `components` module: Perhaps as a followup we can add helper methods there that do partial parsing?

Comment (Manishearth) [reply utils/zerovec/src/varzerovec/slice.rs:94]: something like `components::len_of_buffer` and `components::indices_of_buffer(len)`

File: utils/zerovec/src/varzerovec/slice.rs (hunk 1 above)
Comment (sffc) [reply utils/zerovec/src/varzerovec/slice.rs:71]: According to CI:

Before:
```
-rwxr-xr-x 1 runner docker    59664 Dec 20 02:10 optim5.elf
```

After:
```
-rwxr-xr-x 1 runner docker    58528 Dec 20 03:16 optim5.elf
```

So this PR actually reduces code size overall. I'm happy to investigate future improvements later then.

File: utils/zerovec/src/varzerovec/slice.rs (hunk 2 above)
Comment (sffc) [reply utils/zerovec/src/varzerovec/slice.rs:94]: Yes, that would work for me.

File: utils/zerovec/src/varzerovec/slice.rs (hunk 1 above)
Comment (Manishearth) [reply utils/zerovec/src/varzerovec/slice.rs:71]: Yeah, it reduces codesize by sharing all the codepaths better I think

**Review by sffc — APPROVED** [review]

**Comment by sffc** [comment]
Merging this on behalf of @Manishearth so it doesn't bitrot again

**Pull request merged**